### DISCUSSION TITLE
Stott Robots Handler 3.0

### DISCUSSION THREAD

**Discussion by GeekInTheNorth** [GeekInTheNorth/Stott.Optimizely.RobotsHandler#77] [Announcements]
# Release 3.0

This package has been released to nuget.optimizely.com and nuget.org.

## Features

- #68 
  - Added the ability to define robots content on a host by host basis for any site within an Optimizely CMS instance.
  - A robots.txt configuration with a Host of "Default" will apply it's content to any host if a host specific match is not found.

![Stott Robots Handler Listing](https://github.com/GeekInTheNorth/Stott.Optimizely.RobotsHandler/blob/main/StottRobotsList1.png?raw=true)

## Bugs

- #67 
  - Prevented a 500 error when overriding Cache-Control headers for static files.

## Hotfix 3.0.1

- #79 
  - Corrected Authorization attributes on the Administration Page
    - Please note that while the Administration Page could be reached without a login, none of the APIs were accessible to a user that is not logged in.  This means no-one would be able to view, edit or delete a robots configuration.

<hr /><em>This discussion was created from the release <a href='https://github.com/GeekInTheNorth/Stott.Optimizely.RobotsHandler/releases/tag/v3.0.0'>Stott Robots Handler 3.0.0</a>.</em>

_Help me celebrate this release with a coffee on ko-fi:_

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/V7V0RX2BQ)

**Comment by justinrichwillowtree** [comment]
@GeekInTheNorth have you tried this against any .Net 8 builds?  We're experiencing issues with the tool in our CMS 12 instance on a .Net 8 version.  We think this may have began after our update from .Net 6.

**Comment by GeekInTheNorth** [comment]
Hello @justinrichwillowtree 

I have a .NET 8.0 test platform hosted in Linux on an Azure Web App that I use for testing both Stott Robots Handler and Stott Security before releasing.  Stott Robots Handler is compiled for .NET 6.0 but the backwards compatibility nature of .NET means that you can install a .NET 6 package in .NET 8 and it should run without issue.

If you can elaborate on the issues you're having, I can either advise or look into this for you.

Regards,
Mark

**Reply by justinrichwillowtree** [reply]
When we make changes to the robots.txt and click save, we're seeing a generic error that is not logging or triggering an actual error in logs or console.
![image](https://github.com/user-attachments/assets/89b8a6f9-a867-48ce-b31e-13495664b0fa)

**Comment by GeekInTheNorth** [comment]
Hello @justinrichwillowtree 

When you click the save button in the browser, can you look at the network tab and see what the response is for the following End Point: `/stott.robotshandler/api/save` The Message for the exception thrown will be in the content for the error result.

Can you tell me what that message is and what exact status code is being returned.

If your logging is working correctly, you should get the full exception in the output.  Here is the relevant controller action here:

```
    [HttpPost]
    [Route("/stott.robotshandler/api/[action]")]
    public IActionResult Save(SaveRobotsModel formSubmitModel)
    {
        try
        {
            if (_service.DoesConflictExists(formSubmitModel))
            {
                return new ContentResult
                {
                    StatusCode = (int)HttpStatusCode.Conflict,
                    Content = "A robots configuration already exists for this site and host combination.",
                    ContentType = "text/plain"
                };
            }
            _service.Save(formSubmitModel);

            return new OkResult();
        }
        catch (Exception exception)
        {
            _logger.LogError(exception, "Failed to save robots.txt content for {siteName}", formSubmitModel.SiteName);
            return new ContentResult
            {
                StatusCode = (int)HttpStatusCode.InternalServerError,
                Content = exception.Message,
                ContentType = "text/plain"
            };
        }
    }
```

**Reply by justinrichwillowtree** [reply]
@GeekInTheNorth That got me to the root of it... so your front-end is posting the request to `/api/save/` with the trailing slash and we had implemented a redirect rule on all trailing slashes.  Had your post been to `/api/save` this wouldn't have occurred... the redirect then was a GET vs. POST.  Thanks for taking the time to look into it.  I've added an `ignoreRoutes` array to the initializer and passing your `stott.robotshandler/api/save/` to it.

**Reply by GeekInTheNorth** [reply]
Interestingly I put the slash on because of redirects from other systems that were putting slashes on the end 😂

**Reply by GeekInTheNorth** [reply]
@justinrichwillowtree as a side note, if you decide to use my other module Stott Security, you would need to make the same exclusions. 

I do recommend you check it out as it caters for CSP, CORS and other response headers. https://github.com/GeekInTheNorth/Stott.Security.Optimizely
